src/main/java/org/opentripplanner/framework/lang/PredicateUtils.java

```diff
@@ -0,0 +1,24 @@
+package org.opentripplanner.framework.lang;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Utility for building predicates to be used for filtering streams.
+ */
+public class PredicateUtils {
+
+  /**
+   * Build a predicate that uses the {@code keyExtractor} to remove any key that has already
```

Comment:
You could document if you want to deduplicate by Object identity or Object equality.
The implementation deduplicates by equality (putIfAbsent), but the unit test tests only deduplication by identity, since you are using the same String object and not a different instance `new String("HELLO")`

Reply:
8d39386af110705401a3d4764e072af34e47c945